types/terser-webpack-plugin/index.d.ts

```diff
@@ -23,32 +20,104 @@ declare namespace TerserPlugin {
         extractedComments: any;
     }
 
+    interface FileData {
+        readonly filename: string;
+        readonly basename: string;
+        readonly query: string;
+        readonly hash: string;
+    }
+
     interface ExtractCommentOptions {
-        condition: boolean | string | RegExp | ExtractCommentFn | object;
-        filename?: string | FormatFn;
-        banner?: boolean | string | FormatFn;
+        condition: string | RegExp | ExtractCommentFn;
+        filename?: string | FilenameFn;
+        banner?: false | string | FormatFn;
     }
 
-    type ExtractCommentFn = (node: any, comment: any) => (boolean | object);
+    type ExtractCommentFn = (astNode: any, comment: any) => boolean | object;
 
     type FormatFn = (input: string) => string;
 
+    type FilenameFn = (fileData: FileData) => string;
```

Comment:
According to the [documentation ](https://webpack.js.org/plugins/terser-webpack-plugin/#extractcomments) the filename function takes a string as input.

Reply:
Yes, I know this, but again, they have a vague decription here. The source code suggests string or an object:
https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/test/extractComments-option.test.js#L14-L22
and:
```ts
extractComments: {
          condition: /^\**!|@preserve|@license|@cc_on/i,
          filename: (fileData) => {
            // The "fileData" argument contains object with "filename", "basename", "query" and "hash"
            return `${fileData.filename}.LICENSE.txt${fileData.query}`;
          },
          banner: (licenseFile) => {
            return `License information can be found in ${licenseFile}`;
          },
```
https://webpack.js.org/plugins/terser-webpack-plugin/#object

Reply:
Oh, I did not see that. Thanks for clarifying.